Widgets/CodeWidget.h

```diff
@@ -1,16 +1,95 @@
 #ifndef CODEWIDGET_H
 #define CODEWIDGET_H
 
+#include <QFileDialog>
 #include <QFont>
+#include <QInputDialog>
+#include <QMessageBox>
+#include <QPrinter>
+#include <QTextStream>
 #include <QWidget>
 
 class CodeWidget : public QWidget {
+  Q_OBJECT
+  Q_PROPERTY(QString code READ code WRITE setCode USER true)
+
  public:
   explicit CodeWidget(QWidget* parent);
   ~CodeWidget();
 
+  QString code() const;
+  void setCode(QString);
+  int lineCount();
+
+ public slots:
+  void newSource() {
+    QMessageBox::StandardButton reply;
+    reply =
+        QMessageBox::question(this, tr("New Source"), tr("Are you sure you want to clear the source and start over?"),
+                              QMessageBox::Yes | QMessageBox::No);
+    if (reply == QMessageBox::Yes) {
+      this->setCode("");
+    }
+  }
+
+  void loadSource() {

```

Comment:
Is there a good reason to inline these methods? They're pretty big. If it were a template, I'd understand.

Reply:
Yeah, they are exactly the same for `CodeWidgetPlain.cpp` and `CodeWidgetScintilla.cpp`, hence I didn't want to duplicate code so I just inlined it in the header. I feel that adding `CodeWidget.cpp` to the mix would be confusing, but we can do that if it's really necessary.